addons/pos_coupon/models/coupon_program.py

```diff
@@ -69,15 +69,25 @@ def action_view_pos_orders(self):
 
     @api.depends("rule_products_domain")
     def _compute_valid_product_ids(self):
+        domain_products = {}
         for program in self:
-            domain = ast.literal_eval(program.rule_products_domain) if program.rule_products_domain else []
-            program.valid_product_ids = self.env["product.product"].search(domain).ids
+            product_ids = domain_products.get(program.rule_products_domain)
+            if product_ids is None:
+                domain = ast.literal_eval(program.rule_products_domain) if program.rule_products_domain else []
+                product_ids = self.env["product.product"].search(domain, order="id").ids
```

Comment:
You can do a `search_read`instead of a search and then read the id

Reply:
Both sentence are executing the same query
 - <img width="1511" alt="Screenshot 2024-05-08 at 8 35 38 a m" src="https://github.com/odoo/odoo/assets/6644187/770d6900-400e-4b47-82d9-1fe32021e7f0">

search_read will require an extra process to transform the dict to list of ids

Reply:
Yeah that's good for me